TRegistrationICP Refactoring, Float64, and Robust-Kernel support.

reyanshsolis opened the pull request:
Dependent PR: #3588

Changes:
- [x] Float64 Support.
- [x] Robust Kernel Support.
- [x] Refactoring RegistrationMultiScaleICP.


<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/intel-isl/open3d/3649)
<!-- Reviewable:end -->

stotko reviewed (changes requested):
> Reviewed 39 of 39 files at r1.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3649)* status: all files reviewed, 12 unresolved discussions (waiting on @reyanshsolis and @theNded)

___
*[cpp/open3d/t/geometry/kernel/GeometryMacros.h, line 64 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR3v276FbRVy6QKeIY:-MdR3v28ALYW1m64KHlT:b42tyel) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/geometry/kernel/GeometryMacros.h#L64)):*
> ```objc
> #endif
> 
> #define OPEN3D_ABS(a) (a < 0 ? -a : a)
> ```

Defining such macros can be dangerous and is considered bad practice. For instance `OPEN3D_ABS(x - 1)` for `x < 1` will be substituted to `- x - 1` instead of `- (x - 1)`. Using parentheses like in `OPEN3D_SQUARE` workarounds this problem, but we should avoid macros instead as much as possible.
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransform.h, line 42 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR5Pcq5rJg0I3Gywe3:-MdR5Pcq5rJg0I3Gywe4:bb7462z) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransform.h#L42)):*
> ```objc
> /// \param target_points target points indexed according to correspondences.
> /// \param target_normals target normals indexed according to correspondences.
> /// \param correspondence_indices TODO.
> ```

TODO
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransform.h, line 55 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR5TIFDd5ZWlX0N_69:-MdR5TIFDd5ZWlX0N_6A:bb7462z) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransform.h#L55)):*
> ```objc
> /// \param source_points source points indexed according to correspondences.
> /// \param target_points target points indexed according to correspondences.
> /// \param correspondence_indices TODO.
> ```

TODO
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransform.cpp, line 30 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR5_c4Duw0rxtyWDvm:-MdR5_c4Duw0rxtyWDvn:binosey) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransform.cpp#L30)):*
> ```cpp
> 
> #include "open3d/t/pipelines/kernel/ComputeTransformImpl.h"
> #include "open3d/utility/Timer.h"
> ```

Is this a leftover from profiling?
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransformCPU.cpp, line 54 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR5sGK3aIxsXIYhcgT:-MdR5sGK3aIxsXIYhcgU:b2kjpvx) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransformCPU.cpp#L54)):*
> ```cpp
>         const int n,
>         scalar_t *global_sum,
>         funct_t op) {
> ```

What is `op`? Is it used to compute weights?
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransformCPU.cpp, line 220 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR6AU86rjoD-IaE6At:-MdR6AU86rjoD-IaE6Au:b-ellhcm) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransformCPU.cpp#L220)):*
> ```cpp
> 
>     for (int i = 0; i < 6; i++) {
>         mean_1x7[i] = mean_1x7[i] / mean_1x7[6];
> ```

Is is possible to have `mean_1x7[6] == 0`? If yes, we need some fallback.
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransformCPU.cpp, line 276 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR6N4-BsJjZGQRo3Qh:-MdR6N4-BsJjZGQRo3Qi:b-q64pkt) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransformCPU.cpp#L276)):*
> ```cpp
>     for (int i = 0, j = 0; j < 3; j++) {
>         for (int k = 0; k < 3; k++) {
>             sxy_ptr[j * 3 + k] = sxy_1x9[i++] / mean_1x7[6];
> ```

See above.
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransformCUDA.cu, line 37 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR6VTG-w0g7c6zDra_:-MdR6VTG-w0g7c6zDraa:b-q64pkt) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransformCUDA.cu#L37)):*
> ```Cuda
> #include "open3d/t/pipelines/registration/RobustKernel.h"
> #include "open3d/t/pipelines/registration/RobustKernelImpl.h"
> #include "open3d/utility/Timer.h"
> ```

See above.
___
*[cpp/open3d/t/pipelines/kernel/ComputeTransformCUDA.cu, line 54 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR6Rl3Bp0Wk-MnWkzp:-MdR6Rl3Bp0Wk-MnWkzq:b-q64pkt) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/ComputeTransformCUDA.cu#L54)):*
> ```Cuda
>         const int n,
>         scalar_t *global_sum,
>         funct_t op) {
> ```

See above.
___
*[cpp/open3d/t/pipelines/kernel/Reduction6x6Impl.cuh, line 183 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR6ixuDWgfFLufYa2P:-MdR6ixuDWgfFLufYa2Q:b-6abbt4) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/kernel/Reduction6x6Impl.cuh#L183)):*
> ```Cuda
>         BlockReduceSum<scalar_t, BLOCK_SIZE>(tid, local_sum0, local_sum1);
>         if (tid == 0) {
>             OPEN3D_ATOMIC_ADD(&global_sum[27], local_sum0[0]);
> ```

nit: We can keep `atomicAdd` as long as this function will not become `__host__`-compatible.
___
*[cpp/open3d/t/pipelines/registration/Registration.cpp, line 190 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR7FB_A6rJL3zA8Rrj:-MdR7FB_A6rJL3zA8Rrk:b-jftxid) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/open3d/t/pipelines/registration/Registration.cpp#L190)):*
> ```cpp
>     // ---- Asserts END -------------------------------------------------------
> 
>     // ---- Creating pointcloud pyramid START ---------------------------------
> ```

Such comments usually make me think whether moving all that code out to a dedicated function would make everything cleaner. Would that be an option here (and/or for the others)?
___
*[cpp/tests/t/pipelines/registration/Registration.cpp, line 244 at r1](https://reviewable.io/reviews/intel-isl/open3d/3649#-MdR8Pwn4ndVqjz-J4xE:-MdR8Pwn4ndVqjz-J4xF:bkv9wkp) ([raw file](https://github.com/intel-isl/open3d/blob/f8e4c67a7c7c7f9618f50721ec64549fa6d4903b/cpp/tests/t/pipelines/registration/Registration.cpp#L244)):*
> ```cpp
>                 initial_transform_t,
>                 t_reg::TransformationEstimationPointToPlane(
>                         t_reg::RobustKernel(t_reg::RobustKernelMethod::L1Loss,
> ```

Could we parameterize the test over `RobustKernelMethod` to cover all combinations?


<!-- Sent from Reviewable.io -->

reyanshsolis commented:
Hi @stotko,
I have moved this to PR #3690.
I have added your suggested changes. 
- Renamed `op` to `GetWeightFromRobustKernel`.
- using atomicAdd.
- replaced TODO with the latest definition of correspondences. [Changed from having a pair of tensors to having just one tensor with -1 for no correspondences, as returned by NNS module].  
- Cleaned up `RegistrationMultiScaleICP` function, by using smaller static functions. This will also be helpful if in the future we decide to use a combined kernel for `NNS + ComputeTransform`, or change anything similarly. 
- `mean_1x7[6]` is the count for valid correspondences. So, it is 0 when there are no correspondences. I have added a check in the `Registration` level function to throw an error for that case.
- Added unit tests for Robust Kernel separately.
